src/core/idb-cache.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * A promise based interface to a key-value based indexedDB cache
+ */
+
+const databases = new Map();
+
+export default class IDBCache {
+  /**
+   * @constructor
+   * @param {string} name             name of database
+   * @param {array} stores            name of stores
+   * @param {object} Options
+   *    @param {integer} version      database version
+   *    @param {string} defaultStore  name of default store
+   */
+  constructor(name, stores = [], { version = 1, defaultStore }) {
+    this.name = name;
+    this.stores = typeof stores === "string" ? [stores] : stores;
+    this.version = version;
+    this.defaultStore = defaultStore || this.stores[0];
+    if (!databases.has(name)) {
+      const dbPromise = new Promise((resolve, reject) => {
+        const request = window.indexedDB.open(this.name, this.version);
+        request.onerror = () => reject(request.error);
+        request.onupgradeneeded = () =>
+          this.stores.forEach(storeName =>
+            request.result.createObjectStore(storeName));
+        request.onsuccess = () => resolve(request.result);
+      });
+      databases.set(this.name, dbPromise);
+    }
+  }
+
+  /**
+   * @private returns a promise to get an IDBObjectStore
+   * @param {string} storeName      store name
+   * @param {string} mode           IDBTransaction.mode
+   */
+  async ready(storeName, mode) {
```

Comment:
should i put this promise into a separate function (`getStore`) as you said ready is doing too much work? if we put this into each method (get, set), it'll become redundant

Reply:
I don’t have a strong opinion. I guess we don’t want the caller getting access to the raw store, so maybe just inline or make getStore() a private function to this module.

Reply:
Shouldn't it be 
``` js
async ready(){
  const db = await databases.get(this.name);
  databases.set(this.name, db); // the resolved IDBDatabase, so we can access it in getStore
  return; 
}
```

Reply:
also, shouldn't ready be also private?

Reply:
First question, I respectfully think no because you still need to “tick” on await. So there is no advantage to changing the type held by the map. 

With second question, maybe. It depends if we want a generic means to check if the cache is ready to be used (without doing any actual operation on it). So, just:

```js
const cache = new IDBChace();
const readyPromise = cache.ready();
// do other stuff
readyPromise.then(addThings);
```

All the other operations require some interaction with the database.